### PR TITLE
feat(btc): added checks and tests for btc fee

### PR DESCRIPTION
**Pull request by rwlockbg**

### CODE REVIEW

**Comment by birozuru** [comment]
# eBurger report

**Comment by birozuru** [comment]
# Slither report
**THIS CHECKLIST IS NOT COMPLETE**. Use `--show-ignored-findings` to show all the results.
Summary
 - [arbitrary-send-eth](#arbitrary-send-eth) (1 results) (High)
 - [missing-zero-check](#missing-zero-check) (1 results) (Low)
 - [solc-version](#solc-version) (1 results) (Informational)
 - [reentrancy-unlimited-gas](#reentrancy-unlimited-gas) (1 results) (Informational)
## arbitrary-send-eth
Impact: High
Confidence: Medium
 - [ ] ID-0
[Minting.mint(address,uint256,uint32,bytes,address)](https://github.com/botanix-labs/Macbeth/blob/2dbc608c1d5c6da26af083c73fe44ab1176c36dd/contracts/src/Minting.sol#L45-L84) sends eth to arbitrary user
	Dangerous calls:
	- [address(destination).transfer(amount)](https://github.com/botanix-labs/Macbeth/blob/2dbc608c1d5c6da26af083c73fe44ab1176c36dd/contracts/src/Minting.sol#L80)
	- [address(refundAddress).transfer(txCost)](https://github.com/botanix-labs/Macbeth/blob/2dbc608c1d5c6da26af083c73fe44ab1176c36dd/contracts/src/Minting.sol#L81)

https://github.com/botanix-labs/Macbeth/blob/2dbc608c1d5c6da26af083c73fe44ab1176c36dd/contracts/src/Minting.sol#L45-L84


## missing-zero-check
Impact: Low
Confidence: Medium
 - [ ] ID-1
[Minting.mint(address,uint256,uint32,bytes,address).refundAddress](https://github.com/botanix-labs/Macbeth/blob/2dbc608c1d5c6da26af083c73fe44ab1176c36dd/contracts/src/Minting.sol#L50) lacks a zero-check on :
		- [address(refundAddress).transfer(txCost)](https://github.com/botanix-labs/Macbeth/blob/2dbc608c1d5c6da26af083c73fe44ab1176c36dd/contracts/src/Minting.sol#L81)

https://github.com/botanix-labs/Macbeth/blob/2dbc608c1d5c6da26af083c73fe44ab1176c36dd/contracts/src/Minting.sol#L50


## solc-version
Impact: Informational
Confidence: High
 - [ ] ID-2
Version constraint ^0.8.13 contains known severe issues (https://solidity.readthedocs.io/en/latest/bugs.html)
	- VerbatimInvalidDeduplication
	- FullInlinerNonExpressionSplitArgumentEvaluationOrder
	- MissingSideEffectsOnSelectorAccess
	- StorageWriteRemovalBeforeConditionalTermination
	- AbiReencodingHeadOverflowWithStaticArrayCleanup
	- DirtyBytesArrayToStorage
	- InlineAssemblyMemorySideEffects
	- DataLocationChangeInInternalOverride
	- NestedCalldataArrayAbiReencodingSizeValidation.
It is used by:
	- [^0.8.13](https://github.com/botanix-labs/Macbeth/blob/2dbc608c1d5c6da26af083c73fe44ab1176c36dd/contracts/src/Minting.sol#L2)

https://github.com/botanix-labs/Macbeth/blob/2dbc608c1d5c6da26af083c73fe44ab1176c36dd/contracts/src/Minting.sol#L2


## reentrancy-unlimited-gas
Impact: Informational
Confidence: Medium
 - [ ] ID-3
Reentrancy in [Minting.mint(address,uint256,uint32,bytes,address)](https://github.com/botanix-labs/Macbeth/blob/2dbc608c1d5c6da26af083c73fe44ab1176c36dd/contracts/src/Minting.sol#L45-L84):
	External calls:
	- [address(destination).transfer(amount)](https://github.com/botanix-labs/Macbeth/blob/2dbc608c1d5c6da26af083c73fe44ab1176c36dd/contracts/src/Minting.sol#L80)
	- [address(refundAddress).transfer(txCost)](https://github.com/botanix-labs/Macbeth/blob/2dbc608c1d5c6da26af083c73fe44ab1176c36dd/contracts/src/Minting.sol#L81)
	Event emitted after the call(s):
	- [Mint(destination,amount,bitcoinBlockHeight,metadata)](https://github.com/botanix-labs/Macbeth/blob/2dbc608c1d5c6da26af083c73fe44ab1176c36dd/contracts/src/Minting.sol#L83)

https://github.com/botanix-labs/Macbeth/blob/2dbc608c1d5c6da26af083c73fe44ab1176c36dd/contracts/src/Minting.sol#L45-L84

File: bin/btc-server/src/util.rs [hunk 1]
```diff
@@ -13,8 +13,8 @@ use reth_btc_wallet::psbt::PsbtInputExt;
 use crate::{database, Error};
 
 lazy_static! {
-    // TODO get a fee max amount
-    static ref MAX_AMOUNT: bitcoin::Amount = bitcoin::Amount::from_sat(21_000_000 * 100_000_000);
+    static ref MAX_FEE_AMOUNT: bitcoin::Amount = bitcoin::Amount::from_sat((0.5 * 100_000_000.0) as u64); // 0.5 BTC
```
Comment (arminsabouri) [inline bin/btc-server/src/util.rs]: We need some group thought on this. @PhoenixCodeCraft @jeffgeudens
What should be a the maximum BTC fee any pegout should pay

File: bin/btc-server/src/util.rs [hunk 2]
```diff
@@ -223,19 +223,42 @@ pub fn validate_psbt(
...
-    if fee < Amount::ZERO {
-        return Err(ValidatePSBTError::FeeSanityCheck("Fee cannot be negative"));
+    let fee = match psbt.fee() {
+        Ok(fee) => fee,
+        Err(e) => match e {
+            bitcoin::psbt::Error::NegativeFee => {
+                return Err(ValidatePSBTError::FeeSanityCheck("Fee cannot be negative"))
+            }
+            bitcoin::psbt::Error::FeeOverflow => {
+                return Err(ValidatePSBTError::FeeSanityCheck("Fee oveflow"))
```
Comment (arminsabouri) [inline bin/btc-server/src/util.rs]: lil typo here

File: bin/btc-server/src/util.rs [hunk 3]
```diff
@@ -223,19 +223,42 @@ pub fn validate_psbt(
...
+            }
+            _ => return Err(ValidatePSBTError::FeeCalculationError(e)),
+        },
+    };
+
+    if fee > *MAX_BTC_AMOUNT {
+        return Err(ValidatePSBTError::FeeSanityCheck("Fee cannot be greater than max btc amount"));
     }
```
Comment (arminsabouri) [inline bin/btc-server/src/util.rs]: I think for max amount we want the check to be sum(output.values) <= MAX_BTC_AMOUNT

**Review by arminsabouri — COMMENTED** [review]
CodeACK.
The code looks correct to me! Great tests!
Just want to pull someother people in to talk about the constants being set here

File: bin/btc-server/src/util.rs (hunk 2 above)
Comment (rwlockbg) [reply bin/btc-server/src/util.rs]: fixed

File: bin/btc-server/src/util.rs (hunk 3 above)
Comment (rwlockbg) [reply bin/btc-server/src/util.rs]: done.

File: bin/btc-server/src/util.rs (hunk 1 above)
Comment (arminsabouri) [reply bin/btc-server/src/util.rs]: cc @jeffgeudens @PhoenixCodeCraft

Comment (arminsabouri) [reply bin/btc-server/src/util.rs]: Outcome of slack DMs:
Lets drop of the `MAX_FEE_AMOUNT` and associated tests and code 
@rwlockbg

Comment (rwlockbg) [reply bin/btc-server/src/util.rs]: done, pls re-approve

**Review by scottmillner — DISMISSED** [review]

**Review by arminsabouri — DISMISSED** [review]

**Comment by arminsabouri** [comment]
We need a rebase on this and there is one test that is broken I can also help patch this if you do not find the time

**Pull request merged**